### ISSUE TITLE
 SGAccountController LogOn action method

### ISSUE THREAD

**Issue by leniel** [kahanu/Security-Guard#3]
This action method returns the empty view instead of the model in case of a failed login attempt.

This:

```
// If we got this far, something failed, redisplay form
return RedirectToAction("LogOn");
```

Should be:

```
// If we got this far, something failed, redisplay form
return View(model);
```

This is important so that the errors appear in the view. Today the errors are being hidden.

**Comment by kahanu** [comment]
Leniel, thanks I'll update this soon.  King Wilder

**Comment by kahanu** [comment]
Leniel, this does work as it should.  The RedirectToAction("LogOn") does redirect back to the LogOn Action and then to the LogOn view and display the correct error message.  Have you made changes in your application?  This should all work as designed.

I'm following the PRG (Post/Redirect/Get) pattern, which is better than simply displaying the view.

Thanks.

**Comment by leniel** [comment]
How can this work if you don't pass the model which has the ModelErrors back to the view?

```
@Html.ValidationSummary(true)
```

will show nothing if don't pass the model...

The way it was I couldn't see the model errors in the view if the login failed. After doing the modification I mentioned the errors appear as expected.

I didn't modify anything in my app... that's the default behavior I think.

**Comment by leniel** [comment]
I just noted that the same apllies to:

```
public virtual ActionResult ChangePassword(ChangePasswordViewModel model)
{
   ...
}
```

It should `return View(model);`, otherwise, model errors won't appear... :)

**Comment by kahanu** [comment]
Leniel, I understand what you are saying but this works because of the OnActionExecuted method in the base controller class.  It collects any AddModelError message and displays it in the view based on a PRG pattern.  

If you install the NuGet SecurityGuard package as is, and enter in an incorrect logon username and password, you'll see the correct error message.  You don't need to change the code that is created by the NuGet package.

I hope that helps.

**Comment by leniel** [comment]
This is really strange... it didn't work in my case. Maybe because I'm using MVC 4 beta. Beats me... :) Anyway it's working right now with the change I had to make.

**Comment by kahanu** [comment]
Leniel, I think I understand what you mean.  I tested this in a new MVC 4 application and indeed you are right, BUT, and here's the big BUT, :^), the authentication of the users credentials works, even if it doesn't show anything using the default MVC 4 site that is created for you.

The Default site is implementing a partial view, which is very cool, but the authentication and error response mechanism is handled differently than it is typically.

So to make a long story short, the SGAccountController actions still work as they should and they return the correct error messages.  But to make them work with this new partial view implementation, you'll have to do a little modification to the SGAccount controller.

What I did to test this was go into the _LoginPartial.cshtml file and comment out the Logon ActionLink helper and simply add an ActionLink that points to the SGAccount/Logon action, and it opens a view and it works as it should.

Now that I see this, I'll consider adding similar Json actions to the SGAccount controller but leave the existing actions so it can be used either way.

I'll do some testing this weekend and see if I can get something implemented soon.

Thanks for spotting this.  Before you mentioned it I hadn't had a chance to look at the MVC 4 bits yet.  I still have some catching up to do, but you've helped a lot.  I appreciate it.
